jwst/lib/set_telescope_pointing.py

```diff
@@ -306,39 +306,43 @@ def __str__(self):
 GuideStarPosition.__new__.__defaults__ = ((None,) * 3)
 
 
+def NoneFactory():
+    """Set default of any field in the dataclass to None without mypy getting angry"""
+    return dataclasses.field(default_factory=lambda: None)
+
```

Comment:
This does seem a bit sneaky.  Am I misunderstanding, or is this allowed by mypy only because the NoneFactory itself is not marked with typing on its return value?  PyCharm doesn't like the assignments to `NoneFactory()` because it looks like it does not have a return value.

Why not just modify the typing on the dataclass fields to be `np.array | None`, etc?  I spot checked a few and it looks like mypy has no problem with that.

Reply:
Yes that is the only reason it works.

I tried doing this to all of them, and it leads to several difficult-to-fix issues.  I'll make a branch that shows these issues - it would certainly be nicer it we could actually fix them all, but it looked like a ton of work

Reply:
See here: https://github.com/spacetelescope/jwst/actions/runs/11262041086/job/31316829777?pr=8868
I played around a fair bit before trying to fix these, but I don't think it's very easy.  I had a Slack discussion with @stscieisenhamer about this, and he confirmed that many of the parameters are needed in certain cases but not others, and also that it's not always simple to set a non-None default.  It seemed likely to me that a pretty substantial refactor would be needed to fix this.  What do you think?

Reply:
Maybe we should use the [--no-strict-optional](https://mypy.readthedocs.io/en/stable/command_line.html#no-strict-optional) option?  On your other branch, with this option, it has only one complaint about the `set_telescope_pointing` module.

Reply:
Or not.  I just saw the warning that it is an evil option that should not be used.

Reply:
wow, you weren't kidding, the documentation really calls it evil

Reply:
The other thing we can do is `| Any` instead of `| None`.  

I think the way you're doing it now is unnecessarily confusing, and it looks like it's effectively disabling the useful part of typing.  This one, reported on your other branch, looks like a real code problem that's not caught in this branch:
```
jwst/lib/set_telescope_pointing.py:1047: error: Missing positional arguments "v2_ref", "v3_ref", "v3yangle", "vparity", "crpix1", "crpix2", "cdelt1", "cdelt2", "vertices_idl" in call to "SIAF"  [call-arg]
```

Reply:
Using `Any` also defeats a lot of the utility of type hinting, but I agree in this case it may be better than what I did.  And nice catch on that one error - this is indeed the only one that remains using Any types, which I assume is what you also tried locally.

There is functionally no difference between `type | Any` and simply `Any`, but I would propose to leave the pipes in there because it's more human-readable, i.e., "this should generally be `type` but we needed to allow `Any` to get mypy to shut up.

See the most recent commit where I did as you suggest and fixed the "missing positional arguments" bug

Reply:
Agreed, I like the pipes for readability.  And also agreed that Any isn't super helpful as a type, but better than either a major refactor or just working around typing entirely.

Just one more suggestion here... You many not need `| Any` everywhere -- `| None` might be sufficient for some parameters.  Trying it locally, I just added it to the arrays it was complaining about, until it stopped complaining.

Reply:
done, lmk how it looks or any other comments you have on any of the changes, I know there are a lot

Reply:
Thank you!  I think that looks good now.

I think everything else is fine, pending review from either Brett or Zach, but I'll take another glance through tomorrow, with fresh eyes.